File: server/src/main/com/broll/gainea/server/core/map/Area.java
package com.broll.gainea.server.core.map;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class Area extends Location {

    private Set<Location> adjacentLocations = new HashSet<>();
    private AreaType type;
    private String name;
    private AreaID id;

    public Area(AreaID id) {
        this.id = id;
    }

    public AreaID getId() {
        return id;
    }

    public void setType(AreaType type) {
        this.type = type;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void addAdjacentLocation(Location location) {
        adjacentLocations.add(location);
        if (location instanceof Area) {
            ((Area) location).adjacentLocations.add(this);
        }
    }

    public boolean equals(Object o) {
        if (o instanceof Area) {
            return ((Area) o).getNumber() == getNumber();
        }
        return false;
    }

    @Override
    public String toString() {
        return name;
    }

    public AreaType getType() {
        return type;
    }

    @Override
    public Set<Location> getConnectedLocations() {
        return adjacentLocations;
    }

}

File: server/src/main/com/broll/gainea/server/core/map/Area.kt
package com.broll.gainea.server.core.map

class Area(val id: AreaID, val type: AreaType,
           val name: String, coordinates: Coordinates) : Location(coordinates) {

    fun addAdjacentLocation(location: Location) {
        connectedLocations.add(location)
        if (location is Area) {
            location.connectedLocations.add(this)
        }
    }

    override fun equals(o: Any?): Boolean {
        return if (o is Area) {
            o.number == number
        } else false
    }

    override fun toString(): String {
        return name
    }
}
